hpedockerplugin/hpe/hpe_3par_mediator.py

```diff
@@ -0,0 +1,1032 @@
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+"""HPE 3PAR Mediator for OpenStack Manila.
+This 'mediator' de-couples the 3PAR focused client from the OpenStack focused
+driver.
+"""
+import sh
+import six
+
+from oslo_log import log
+from oslo_service import loopingcall
+from oslo_utils import importutils
+from oslo_utils import units
+
+from hpedockerplugin import exception
+from hpedockerplugin.i18n import _
+from hpedockerplugin import fileutil
+
+hpe3parclient = importutils.try_import("hpe3parclient")
+if hpe3parclient:
+    from hpe3parclient import file_client
+
+LOG = log.getLogger(__name__)
+MIN_CLIENT_VERSION = (4, 0, 0)
+DENY = '-'
+ALLOW = '+'
+FULL = 1
+THIN = 2
+DEDUPE = 6
+ENABLED = 1
+DISABLED = 2
+CACHE = 'cache'
+CONTINUOUS_AVAIL = 'continuous_avail'
+ACCESS_BASED_ENUM = 'access_based_enum'
+SMB_EXTRA_SPECS_MAP = {
+    CACHE: CACHE,
+    CONTINUOUS_AVAIL: 'ca',
+    ACCESS_BASED_ENUM: 'abe',
+}
+IP_ALREADY_EXISTS = 'IP address %s already exists'
+USER_ALREADY_EXISTS = '"allow" permission already exists for "%s"'
+DOES_NOT_EXIST = 'does not exist, cannot'
+LOCAL_IP = '127.0.0.1'
+LOCAL_IP_RO = '127.0.0.2'
+SUPER_SHARE = 'DOCKER_SUPER_SHARE'
+TMP_RO_SNAP_EXPORT = "Temp RO snapshot export as source for creating RW share."
+
+
+class HPE3ParMediator(object):
+    """3PAR client-facing code for the 3PAR driver.
+    Version history:
+        1.0.0 - Begin Liberty development (post-Kilo)
+        1.0.1 - Report thin/dedup/hp_flash_cache capabilities
+        1.0.2 - Add share server/share network support
+        1.0.3 - Use hp3par prefix for share types and capabilities
+        2.0.0 - Rebranded HP to HPE
+        2.0.1 - Add access_level (e.g. read-only support)
+        2.0.2 - Add extend/shrink
+        2.0.3 - Fix SMB read-only access (added in 2.0.1)
+        2.0.4 - Remove file tree on delete when using nested shares #1538800
+        2.0.5 - Reduce the fsquota by share size
+                when a share is deleted #1582931
+        2.0.6 - Read-write share from snapshot (using driver mount and copy)
+        2.0.7 - Add update_access support
+        2.0.8 - Multi pools support per backend
+        2.0.9 - Fix get_vfs() to correctly validate conf IP addresses at
+                boot up #1621016
+    """
+
+    VERSION = "2.0.9"
+
+    def __init__(self, host_config, config):
+        self._host_config = host_config
+        self._config = config
+        self._client = None
+        self.client_version = None
+
+    @staticmethod
+    def no_client():
+        return hpe3parclient is None
+
+    def do_setup(self, timeout=30):
+
+        if self.no_client():
```

Comment:
@imran-ansari this check might not be required please check.

Reply:
I think this check will be helpful in case user misses out on installing HPE 3PAR Client.

Reply:
As part of plugin, Client is part of it. It is specified in requirements file hence it will download the required modules. Still do we need to have this check?